pkg/dispers/target.go

```diff
@@ -0,0 +1,146 @@
+package enclave
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/cozy/cozy-stack/pkg/dispers/dispers"
+)
+
+func buildQuery(instance dispers.Instance, localQuery dispers.LocalQuery) dispers.Query {
+	// TODO : encrypt outputs
+	query := dispers.Query{
+		Domain:      instance.Domain,
+		LocalQuery:  localQuery,
+		TokenBearer: instance.Token.TokenBearer,
+	}
+
+	return query
+}
+
+func removeIndex(s *[]dispers.Instance, index int) []dispers.Instance {
+	return append((*s)[:index], (*s)[index+1:]...)
+}
+
+// cleanTargetsList clean Targets list inplace
+func cleanTargetsList(list *[]dispers.Instance) error {
+
+	i := 0
+	for i < len(*list) {
+		if isInstInvalid(&(*list)[i]) {
+			// Unvalid item
+			*list = removeIndex(list, i)
+			i--
+		} else {
+			// Valid item
+			// Check for duplicates
+			j := i + 1
+			iHasBeenDeleted := false
+			for !iHasBeenDeleted && j < len(*list) {
+				if (*list)[i].Domain == (*list)[j].Domain {
+					if (*list)[i].SubscriptionDate.After((*list)[j].SubscriptionDate) {
+						// Item i is kept
+						*list = removeIndex(list, j)
+						j--
+					} else {
+						// Item j is kept
+						*list = removeIndex(list, i)
+						iHasBeenDeleted = true
+					}
+				}
+				j++
+			}
+		}
+		i++
+	}
+	// Check missing data
+	return nil
+}
+
+func decryptInputT(in *dispers.InputT) error {
+	return nil
+}
+
+func retrieveData(in *dispers.InputT, queries *[]dispers.Query) ([]map[string]interface{}, error) {
+
+	var data []map[string]interface{}
+	var rowsData []map[string]interface{}
+
+	for _, query := range *queries {
+
+		url := strings.Join([]string{"http:/", query.Domain, "data/_find"}, "/")
+		marshalFindRequest, err := json.Marshal(query.LocalQuery.FindRequest)
+		if err != nil {
+			return nil, nil
+		}
+		resp, err := http.Post(url, "application/json", bytes.NewReader(marshalFindRequest))
+		if err != nil {
+			return nil, nil
+		}
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return nil, nil
+		}
+		err = json.Unmarshal(body, &rowsData)
+		if err != nil {
+			return nil, nil
+		}
+
+		data = append(data, rowsData...)
```

Comment:
If I understand correctly, the Target directly query the stack? Shouldn't you pass through the conductor?